google-ads-examples/src/main/java/com/google/ads/googleads/examples/reporting/ParallelReportDownload.java

```diff
@@ -0,0 +1,249 @@
+        List<ListenableFuture<ReportSummary>> futures = new ArrayList<>();
+
+        // Calls the API for each customer ID on the current report.
+        for (Long customerId : customerIds) {
+          // Uses the gRPC asynchronous API to download the reports in parallel. This saves having
+          // to create/manage our own thread pool.
+          ResponseCountingObserver responseObserver = new ResponseCountingObserver(customerId);
+
+          // Starts the report download in a background thread.
+          serviceClient
+              .searchStreamCallable()
+              .call(
+                  SearchGoogleAdsStreamRequest.newBuilder()
+                      .setCustomerId(customerId.toString())
+                      .setQuery(gaqlQuery)
+                      .build(),
+                  responseObserver);
+
+          // Stores a future to retrieve the results.
+          futures.add(responseObserver.asFuture());
+        }
+
+        // Waits for all pending requests to the current set of customer IDs to complete.
+        //
+        // This is a naive implementation for illustrative purposes. It is possible to optimize the
+        // utilization of each customer ID by providing a queue of work (or similar). However, this
+        // would complicate the example code and so is omitted here.
+        List<ReportSummary> results =
+            Futures.allAsList(futures)
+                .get(DEADLINE_MILLIS - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+
+        System.out.println("Report results for query: " + gaqlQuery);
+        results.forEach(System.out::println);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException | TimeoutException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /** An observer which records a simple count of the result rows received. */
+  private static class ResponseCountingObserver
+      implements ResponseObserver<SearchGoogleAdsStreamResponse> {
+
+    private final long customerId;
+    private final SettableFuture<ReportSummary> future = SettableFuture.create();
+    private AtomicLong numRecords = new AtomicLong(0);
+
+    ResponseCountingObserver(long customerId) {
+      this.customerId = customerId;
+    }
+
+    @Override
+    public void onStart(StreamController controller) {
+      // Nothing to do here.
+    }
+
+    @Override
+    public void onResponse(SearchGoogleAdsStreamResponse response) {
+      // Does something useful with the row. In this case we just count the
+      // records. Note that this method may be called from multiple threads, though always in
+      // sequence.
+      numRecords.incrementAndGet();
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      // Notify that this report failed.
+      notifyComplete(new ReportSummary(customerId, numRecords.get(), t));
```

Comment:
This contradicts to my intuition a bit.
When we say complete, I feel it shouldn't have any errors.
Can't we have another method named notifyError? Or can we name this method more neutral?

Reply:
Updated, PTAL.